pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java

```diff
@@ -793,6 +806,44 @@ private CompletableFuture<Void> filterAndUnloadMatchedNamespaceAsync(String clus
             if (CollectionUtils.isEmpty(shouldUnloadNamespaces)) {
                 return CompletableFuture.completedFuture(null);
             }
+            // If unload type is 'changed', we need to figure out a further subset of namespaces whose placement might
+            // actually have been changed.
+
+            log.debug("Old policy: {} ; new policy: {}", oldPolicy, policyData);
+            if (oldPolicy != null && NamespaceIsolationPolicyUnloadScope.changed.equals(policyData.getUnloadScope())) {
+                // We also compare that the previous primary broker list is same as current, in case all namespaces need
+                // to be placed again anyway.
+                List<String> oldBrokers = new ArrayList<>(policyData.getPrimary());
```

Comment:
```suggestion
                List<String> oldBrokers = new ArrayList<>(oldPolicy.getPrimary());
```

It should be the primary of the old policy.

Reply:
Whoops. Thank you for catching this. Accepted the change.

Reply:
@grssam any chance to add a test to cover that part?

Reply:
@lhotari added. Will need approval again i guess.